client/service/src/lib.rs

```diff
@@ -112,15 +116,20 @@ impl RpcHandlers {
 		&self,
 		json_query: &str,
 	) -> Result<(String, mpsc::UnboundedReceiver<String>), JsonRpseeError> {
-		self.0
+		self.rpc_module
 			.raw_json_request(json_query)
 			.await
 			.map(|(method_res, recv)| (method_res.result, recv))
 	}
 
 	/// Provides access to the underlying `RpcModule`
 	pub fn handle(&self) -> Arc<RpcModule<()>> {
-		self.0.clone()
+		self.rpc_module.clone()
+	}
+
+	/// Provides access to listen addresses
+	pub fn listen_addresses(&self) -> Arc<Vec<Multiaddr>> {
```

Comment:
nit: I'd prefer an `&[Multiaddr]` as return type, we don't want to edit the `Arc` content nor is there a locking mechanism to it.

Reply:
Immutability was my intention. My understanding from the [Arc docs](https://doc.rust-lang.org/std/sync/struct.Arc.html) was that it would be immutable so long as I didn't wrap it in a Mutex/AtomicRef, but I might have misunderstood the semantics.

Reply:
Returning `Arc` here incurs the overhead of an extra clone, but does not keep a borrow on `&self`.

Yes, Arc is an immutable reference counted pointer.

Reply:
If the caller needs the clone, they shall make a copy of the state, but commonly it's only short lived, where `&[X]` is easier to handle.

An `Arc::get_mut`/`unsafe Arc::get_mut_unchecked` would allow you to _try_ to modify it, which is a footgun - it compiles, but always fails at runtime.

Reply:
> Based on that, I think the only difference is if I return a &[Multiaddr], the caller can force the clone to occur only if they actually need a clone?

One has to distinguish between "cloning the `Arc`" and "cloning the `Vec<Multiaddr>`, i.e. the inner value". The former is just an atomic increment of the refcount, where as the latter is expensive.

Returning `&[Multiaddr]` means returning an immutable slice, it means that the vector is never copied (or even visible -- the caller doesn't even know that `Arc` or `Vec` are involved). Since it returns a reference, it borrows `&self` though.

Reply:
```rust
	pub fn listen_addresses(&self) -> &[Multiaddr] {
```

is the most idiomatic to me, at first glance. It will also be among the fastest, as it will boil down to a single fat pointer lookup at most, if I am not mistaken.

To illustrate, due to lifetime elision, the "actual code" there would be something like

```rust
	pub fn listen_addresses<'a>(&'a self) -> &'a [Multiaddr] {
```

Reply:
Notice that the interface has changed, we've taken away the capability of the user to make cheap reference counted copies of the address list, and have hidden the implementation details. That's good.

Now that we've gone that far, by the way, we should check whether we need the `Arc` at all!

Reply:
Depends how often we `clone` the entire structure, which currently is only two atomic ops

Reply:
Never as it appears.